torch/_dynamo/decorators.py

```diff
@@ -437,19 +437,28 @@ class directly; instead, use :func:`mark_dynamic`.
 
 
 @forbid_in_graph
-def mark_unbacked(t, index):
+def mark_unbacked(t, index, strict=False):
     """
     Mark a tensor as having an unbacked dim.  This changes the semantics of operations,
     we will always report the size does not equal zero/one, we will turn asserts
     on this index into runtime asserts, and if you try to get the real value we will
     raise an exception.  In other words, we will treat this dimension as if it was
     data dependent (we do not know anything about its value.)
+
+    For historical reasons, by default if an unbacked dim is specialized, we will
+    happily specialize it and continue. If you want to error in these cases, pass
+    strict=True.
     """
     # You could have copied the mark_dynamic behavior but I'm not convinced
     # it's what you want
     assert not is_traceable_wrapper_subclass(t), "not implemented yet"
 
     if isinstance(index, int):
+        if strict:
+            if not hasattr(t, "_dynamo_strict_unbacked_indices"):
+                t._dynamo_strict_unbacked_indices = set()
+            t._dynamo_strict_unbacked_indices.add(index)
+
         if not hasattr(t, "_dynamo_unbacked_indices"):
```

Comment:
do we need else

Reply:
I don't think so?

Reply:
ah you want to add both in _dynamo_strict_unbacked_indices and _dynamo_unbacked_indices?

Reply:
Yeah the alternative is to change it so instead of a set of ints, we have a set of objects, but I figured it's cleaner to stay consistent with the other `indices` and keep the type as a list of sets of ints